lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/Terms.scala

```diff
@@ -16,82 +17,138 @@ object Terms {
   sealed abstract class DECLARATION {
     def name: String
     def toStr: Coeval[String]
+    def deepCopy: Eval[DECLARATION]
     override def toString: String = toStr()
     def isItFailed: Boolean       = false
   }
   case class FAILED_DEC() extends DECLARATION {
     def name                         = "NO_NAME"
     def toStr: Coeval[String]        = Coeval.now("Error")
     override def isItFailed: Boolean = true
+
+    override def deepCopy: Eval[DECLARATION] =
+      Eval.now(this)
   }
-  case class LET(name: String, value: EXPR) extends DECLARATION {
+  case class LET(name: String, var value: EXPR) extends DECLARATION {
     def toStr: Coeval[String] =
       for {
         e <- value.toStr
       } yield "LET(" ++ name ++ "," ++ e ++ ")"
+
+    override def deepCopy: Eval[LET] =
+      value.deepCopy.map(LET(name, _))
   }
-  case class FUNC(name: String, args: List[String], body: EXPR) extends DECLARATION {
+  case class FUNC(name: String, args: List[String], var body: EXPR) extends DECLARATION {
```

Comment:
Is a body mutable? Can it be called twice?

Reply:
Fixed in https://github.com/wavesplatform/Waves/pull/3092